curator-x-async/src/main/java/org/apache/curator/x/async/modeled/details/ModeledCacheImpl.java

```diff
@@ -188,9 +188,18 @@ private void internalChildEvent(TreeCacheEvent event)
         {
             ZPath path = ZPath.parse(event.getData().getPath());
             Entry<T> entry = entries.remove(path);
-            T model = (entry != null) ? entry.model : serializer.deserialize(event.getData().getData());
-            Stat stat = (entry != null) ? entry.stat : event.getData().getStat();
-            accept(ModeledCacheListener.Type.NODE_REMOVED, path, stat, model);
+            T model = null;
+            if (entry != null) {
+                model = entry.model;
+            }
+            else if (event.getData().getData() != null) {
+                model = serializer.deserialize(event.getData().getData());
+            }
+            if (model != null) {
+                Stat stat = (entry != null) ? entry.stat : event.getData().getStat();
+                accept(ModeledCacheListener.Type.NODE_REMOVED, path, stat, model);
```

Comment:
This changes the behavior though. Is this correct? Previously, `accept(ModeledCacheListener.Type.NODE_REMOVED...` would always be called.

Reply:
You are correct, this change would alter the behavior in that `accept(ModeledCacheListener.NODE_REMOVED... ` would not be called if the node in question doesn't contain any data (ie, a valid Model).

However, I'm not sure this is a bad thing.  If you look above in the code at the implementation for the NODE_ADDED and NODE_UPDATE case, the `accept()` method is also not called if the ZNode in question does not contain any data.

Unit testing doesn't seem to cover this use case either way (the tests all passed without any modification).

Reply:
> Previously, `accept(ModeledCacheListener.Type.NODE_REMOVED...` would always be called.

It is only true for `ModelSerializer.raw` which I think is a unusual usage of modeled framework. With `JacksonModelSerializer`, the `deserialize` will fail hence `ModeledCacheListener.accept` was not called.

I think it is a chance to unify the two. I prefer the second option as there is no sense to accept a `null` model (`raw` or `json`) for intermediate nodes in all event types. Consistency  should play more importance here than potential breaking of `ModelSerializer.raw`.

@Randgalt What do you think ?

Reply:
I might help to mention the use case.

Imagine that we have a model called Widget.

In our application we group Widgets under sub-groups, and we can have many Widgets (100k+).  

As such, we structured ZooKeeper as follows:
/widget/<sub-group A>/<widget 1>
/widget/<sub-group B>/<widget 2>

We want to use ModeledFramework to deserialize widgets, however if we encounter a sub-group ZNode that is part of the ZNode, we'd like to just ignore that rather than cause serialization exceptions.

I will have a look at ModelSerializer.raw and where it fits in here.

Reply:
It appears that ModelSerializer.raw is package private an unused. Can you kindly describe its purpose?

Reply:
`ModelSerializer.raw` is a interface variable which means that it is `public`. Before this pr, the behavior is that:
* With `ModelSerializer.raw` or maybe other serializers that tolerate `null` input, `ModeledCacheListener.accept` will be called with `null` model.
* With `JacksonModelSerializer` or maybe other serializers that do not tolerate `null` input, `ModeledCacheListener.handleException` will be called with exception from `ModelSerializer.deserialize`.

To keep consistent with `NODE_ADDED` and `NODE_UPDATED`, I think all these methods `ModelSerializer.deserialize` and `ModeledCacheListener.accept/handleException` should not be called for intermediate nodes in case of `NODE_REMOVED`. `ModelSerializer` is an open interface, we should provide consistent behavior among `ModelSerializer.raw`, `JacksonModelSerializer` and possible others. It would be helpful if we can assert this in test.

cc @Randgalt @Ryan0751

Reply:
Thank you for the explanation of ModelSerializer.raw (I totally missed that it was defined in an interface).

As for your second suggestion (not calling deserialized and accept/handleException for intermediate nodes in the case of NODE_REMOVED), does that present a bigger risk of breaking existing users if they expect those nodes to contain valid models?

Reply:
"intermediate nodes" I means nodes with `null` bytes, e.g. `sub` in your test. The current fix is acceptable to me.